오픈마켓 [STEP 2] 토니, 앨리

oksunwoo opened the pull request:
안녕하세요, 라이언(@ryan-son ) 🦁!

토니, 앨리 입니다!

말씀드렸던 기한보다 늦게 PR 드립니다..🥲

STEP2 진행하는 과정에서 수많은 시행착오가 있었지만 드디어 제출을 하게되네요..

부족한 부분이 많겠지만 리뷰 잘 부탁드립니다!😊🙏🏻

## ❗️ 고민했던 점

### TableView vs CollectionView?

이번 Step의 예시를 보면 `ListView`는 `TableView`의 모양을 갖고 있는 것으로 확인해볼 수 있었습니다. 처음에는 단순하게 `TableView`를 사용하여 구현을 하면 되겠다고 생각했습니다. 그 과정에서 `CollectionView`를 이용해서도 `TableView` 를 구성해볼 수 있지않을까라고 고민을 해보았고, `CollectionView` 를 사용하는 프로젝트가 이번이 처음이다 보니 많은 경험을 하고 싶어서 `CollectionView`로 구성을 해보았습니다. 또한, `segmentedControl`을 활용에 있어, 어떻게 구현할 수 있을 지 선택의 폭이 넓어질 수 있다고 생각을 해서 `CollectionView`로 구현을 해보았습니다. 😆

### segmentedControl을 통한 뷰 전환

처음 `segmentedControl`을 사용해보다보니, 스위치 전환에 따른 뷰를 어떻게 구성할지 고민을 많이 했습니다. 🧐

뷰를 보여주는 방식에 대해 두가지 방법을 고민해보았습니다!

1. **2개의 view를 한 화면에 동시에 띄우고, isHidden으로 한 view를 감추는 방식**
2. **view를 하나만 사용하고, 화면전환시 cell의 레이아웃들을 변경하는 방식**

이 두가지의 방법 중 무엇이 정답이다는 없겠지만, 각각의 방법들을 선택했을 때의 장,단점들을 각각 고민해보고 비교해보았습니다.

우선, `첫번째 방식`은 사용자가 볼때는 화면이 한개인것처럼 보이지만, 실제로는 두개의 뷰가 그려지고 있기때문에, 메모리적인 측면에서 낭비가 있을것이라고 생각했습니다. 하지만, 개발단계에서 간편하다는 장점이 있고, 이것이 추후에는 유지보수의 편리함으로 다가올 것 같았습니다. 또한 스위치를 수시로 변경할때, 이미 그려진 `view`가 나타나기에 `view`의 전환 시 속도나 성능이 우수할 것이라고 생각했습니다.

`두번째 방식`은 메모리를 아낄 수 있다는 장점을 분명하게 갖고있다고 생각했습니다. 하지만, 코드를 구현하는 과정이나 추후의 유지보수적인 측면에서 굉장히 까다로울 수 있다는 점이 단점으로 다가왔습니다. 그리고 결정적으로는, `segmentControl`을 이용하여 `view`를 전환해줄 때마다 `cell`들의 레이아웃을 바꿔주기때문에 속도나 성능적인 측면에서 떨어진다고 생각했습니다. 저희는 메모리를 조금은 더 쓰더라도, 결국 사용자 경험이 중요시되는것이 더 나은 방법이라고 생각했기 때문에, **첫번째 방법**으로 구현하게 되었습니다. 😙

### defaultContentConfiguration() 사용에 대한 고민

default 스타일의 cell을 사용하면 custom이 제한된다는 생각이 들었습니다. 화면에 띄워줘야하는 상품의 정보는 4가지(이름, 가격, 할인된 가격, 재고수량)인데, default 스타일로는 2가지 정도만 표현할 수 있는 것 같았고, image의 크기를 원하는대로 조정할 수 없다는 점 때문에 customCell을 사용하는 방향으로 진행했습니다.

아래와 같이 image 최대사이즈를 제한해 줄 수는 있는 것 같은데, image 사이즈를 정해주는 다른 방법이 있는지 궁금합니다! 🤔

```swift
content.imageProperties.maximumSize = CGSize(width: 50 , height: 50)
```

## 💢 해결이 되지 않은 점

### ListCell의 실종사건👮🏻

완성단계에서 이것저것 실험을 해보다가 이상한 점을 발견했습니다. 실제로 주어진 요구사항은 셀을 20개까지만 띄우는 거 였지만, 궁금해서 100개 정도 띄워놓고 해보았습니다. 그랬더니 화면에서 다음 사진처럼 갑자기 빈 부분이 나오기 시작했습니다. debug view hierarchy를 띄워놓고 다음과 같이 빈 부분을 클릭해보았더니, 뿌리가 되는 `ListCollectionView` 자체가 클릭이 되는 것을 볼 수가 있었습니다. 사진을 보면, ListCell 자체가 생성이 되어있지 않은 모습을 볼 수가 있는데, 원인을 알아내지 못했습니다... 그리고 또 스크롤을 하면 세번째 사진처럼 어느순간 cell이 다시 생성되어 정상적으로 작동하는 모습을 발견했습니다. 헌데, 같은 방식으로 구현해준 GridView에서는 아무런 문제가 없었습니다...😢 레이아웃의 문제라면 화면에 보여지는 셀들의 모양도 문제가 있어야한다고 생각하는데 어느부분에서 문제가 생긴 것인지 궁금합니다..

<img width="876" alt="스크린샷 2022-01-20 오후 5 54 40" src="https://user-images.githubusercontent.com/55880539/150306021-ea594117-1f92-42af-865b-07f2d5708471.png">

### AccessoryView의 실종사건👮🏻‍♀️

셀의 재사용을 위해서, prepareForReuse 메서드를 구현해주었습니다. 잘 작동한다고 안심하던 찰나, accessoryImageView가 실종된 것을 발견하고, 원인파악을 위해 수사팀을 꾸렸습니다. 하지만, 미제사건으로 남아서...🥷🏻 수사의뢰를 부탁드려도 될까요..? 

현재 저희의 코드는 하이라이트된 코드 부분을 지운 상태입니다.

![image](https://user-images.githubusercontent.com/55880539/150305845-12ae7c99-1648-4464-b1d0-010d8d2c537d.png)


## ❓ 질문 드리고 싶은 부분

### Cell을 재사용할 때, attributedText가 reset되지 않는 문제

customCell 내부에서 `prepareForReuse()` 메소드로 셀에 들어가는 요소들을 nil로 세팅해주었습니다. 그런데 실행했을 때, discountedPriceLabel이 없는 상품이 priceLabel의 attributedText로 있는 취소선이 생기는 현상을 확인했습니다.

1 처음 실행했을 때 / 2 스크롤 내렸다가 올렸을 때 (취소선 생김)
![image](https://user-images.githubusercontent.com/55880539/150306205-adca5fef-ab8a-439a-a204-68ba6c248dff.png)

![image](https://user-images.githubusercontent.com/55880539/150306333-88c58a1b-539a-4840-8031-ad6835732a0b.png)

```swift
// 현재코드
override func prepareForReuse() {
    super.prepareForReuse()   
    ...

    priceLabel.attributedText = nil
    priceLabel.text = nil
    
    ...
}

// 순서가 바뀌면 취소선이 사라지지 않음.
override func prepareForReuse() {
    super.prepareForReuse()
    ...    
    
    priceLabel.text = nil
    priceLabel.attributedText = nil

    ...
}
```

UILabel text의 공식문서를 확인해보면 text의 value에 새로운 값을 할당하면 attributedText의 value도 새로운 값으로 변경된다고 이해했는데, 실제로 `prepareForReuse()`에 `label.text = nil` 만 처리하면 해결이 되지 않았습니다. 추가적으로 `label.attributedText = nil` 과 `label.text = nil` 의 순서도 결과에 영향을 미치는 것을 확인했습니다.

text에 먼저 nil을 할당하고 attribute에 nil을 할당하면 결과가 변하지 않았고, 반대로 attribute에 nil을 먼저 할당하고 text에 nil을 할당하면 저희가 원하는 결과로 바뀌는 것을 확인했습니다.

text와 attributedText 두 속성 모두 nil을 할당해주어야만 하는지와, nil을 할당하는 순서가 결과에 영향을 미치는 이유가 궁금합니다..🥲


### Cell의 이중등록?

처음에 cell을 등록할 때, register 메소드를 사용했는데 등록이 안되어서 뷰가 그려지지 않았습니다. xib나 코드로 만들어진 CollectionViewCell은 반드시 register를 해주어야한다고 알고 있는데 왜 안되는 것인지 궁금합니다! 이중등록이 되었나라는 생각이 들었는데, `register` 메소드 공식문서를 보면 아래와같이 설명되어있고, 동일한 reuse identifier로 class나 nib 파일을 등록한 경우에는 cellClass 매개변수에 지정한 클래스가 이전 항목을 대체한다고 나와있습니다만, 무슨 말인지 잘 이해가 가지 않습니다..ㅠㅠ 일단은 리팩토링 과정에서 `cellRegistration`의 방식을 택하면서 문제는 해결되었지만 이전방식의 문제의 원인이 궁금해서 질문드려요! 🧐

![image](https://user-images.githubusercontent.com/55880539/150306582-7ed129a3-9664-4dd9-9e21-1dd96d252c9e.png)

[Apple Developer Documentation](https://developer.apple.com/documentation/uikit/uitableview/1614888-register)

```swift
// 이전에 사용하려던 방식
func setUpCell() {
    listCollectionView.register(ListCollectionViewCell.self, forCellWithReuseIdentifier: "cell")
        
    dataSource = UICollectionViewDiffableDataSource<Section, ProductInformation>(collectionView: listCollectionView, cellProvider: { (collectionView, indexPath, product) -> ListCollectionViewCell in
            
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as? ListCollectionViewCell else {
            return ListCollectionViewCell()
        }

        cell.setUpImage(url: product.thumbnail)
        cell.setName(name: product.name)
        return cell
    })
}
```

### 레이아웃 관련 로그

오토레이아웃 관련해서 보라색 에러는 뜨지 않지만, 프로젝트를 실행해서 뷰를 스크롤할 때마다 아래와같은 로그가 찍히고 있습니다..! 현재 저희 레이아웃에 문제가 있어서 수정해주어야한다는 뜻일까요..?🥲

![image](https://user-images.githubusercontent.com/55880539/150306616-a9df2e32-a558-43a2-aed0-3c2d37cd3bd8.png)

Allie-e committed:
refactor: dataTask escaping closure 내부 코드 순서 변경 #6

oksunwoo committed:
refactor: 역할 분리를 위한 URLManager 타입 구현 #6

Allie-e committed:
refactor: 하나의 함수가 하나의 일만 하도록 URLRequest 별도의 함수로 구현 #6

oksunwoo committed:
refactor: 메서드 내부 에러의 표시방법 수정 #6

Allie-e committed:
refactor: Parser 타입 decode 메소드 에러처리 수정 #6

oksunwoo committed:
feat: SegmentedControl 구현 #5

- NavigationController 추가
- SegmentedControl 클래스로 UI요소 업데이트

Allie-e committed:
feat: collectionView 생성 #5

- segmentedControl을 이용한 화면전환 구현 및 테스트

oksunwoo committed:
feat: listCollectionView의 Cell타입 구현 #5

Allie-e committed:
feat: Api 통신을 위한 메소드 구현 #5

oksunwoo committed:
feat: cell을 set하는 메서드와 컬렉션뷰와 데이터소스 연결 #5

Allie-e committed:
feat: snapshot을 이용한 UI update test #5

oksunwoo committed:
refactor: customCell을 코드로 구현하는 방식으로 변경 #6

Allie-e committed:
feat: setUpProductImageView 메소드 추가 #5

oksunwoo committed:
feat: labelStackView 레이아웃 구성 #5

Allie-e committed:
feat: setUpPriceStackView 레이아웃 구성 #5

oksunwoo committed:
feat: accessoryImageView 레이아웃 구성 #5

Allie-e committed:
feat: cellRegistration 및 data 전달 #5

Allie-e committed:
refactor: cell 내부 요소들의 auto layout 개선 #6

oksunwoo committed:
refactor: List셀의 레이아웃 수정 #6

Allie-e committed:
feat: cell label 및 imageView에 data 전달 #6

oksunwoo committed:
feat: 셀의 재사용을 위한 prepareForReuse 메서드 구현 #5

Allie-e committed:
refactor: 상품가격 취소선과 가격표시형식 extension 추가 #6

Allie-e committed:
refactor: cell bottom border 추가 및 extension 파일분리 #6

Allie-e committed:
refactor: prepareForReuse 메소드 accessory 초기화 삭제 #6

oksunwoo committed:
feat: Grid뷰의 Cell 구성 #5

Allie-e committed:
refactor: ListCell underline 수정 #6

oksunwoo committed:
feat: grid셀의 imageView 레이아웃 설정 #5

Allie-e committed:
feat: gridCell registration 및 cell layout 설정 #5

oksunwoo committed:
feat: labelStackView의 레이아웃 설정 #5

Allie-e committed:
feat: priceStackView 구현 및 레이아웃 설정 #5

oksunwoo committed:
feat: Grid뷰 셀에 알맞은 형식으로 데이터 전달 #5

Allie-e committed:
feat: grid cell border 추가 #5

oksunwoo committed:
refactor: cell간 상하 간격을 변경 #6

Allie-e committed:
refactor: Cell 내부 메소드 네이밍 수정 #6

oksunwoo committed:
refactor: 접근제어자 설정 #6

Allie-e committed:
refactor: price label attributtedText 초기화 추가 #6

oksunwoo committed:
chore: 파일 및 구조 변경 #6

ryan-son commented on OpenMarket/OpenMarket/Base.lproj/Main.storyboard:
```diff
@@ -1,24 +1,115 @@
+                                            <autoresizingMask key="autoresizingMask"/>
+                                        </collectionViewCellContentView>
+                                    </collectionViewCell>
+                                </cells>
```
이 프로토타입 셀은 활용하지 않는다면 삭제해도 좋지 않을까요?

Allie-e replied:
기존에 스토리보드로 작업할 때 생성된 셀을 지우는 걸 깜빡했습니다ㅎㅎ 삭제해주었습니다!

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+import UIKit
+
+@available(iOS 14.0, *)
```
Declaration attributes 중 available 속성을 사용해보셨네요! 👏 컴파일러가 이 속성을 명시하라고 한 이유는 무엇일까요?
앱 배포 타겟 OS 버전(iOS 13.2)과 활용하신 메서드의 사용가능 OS 버전(14.0)에 차이가 있다면 그 사이의 OS를 사용하고 있는 사람들을 위한 분기처리가 필요할 것 같아요!

oksunwoo replied:
우선, 컴파일러가 이 속성을 명시하라고 한 이유는 앱의 배포버전이 낮고 개발자가 사용하는 기능은 해당앱의 버전보다 높은 OS에서 지원하는 경우, 해당기능은 높은 OS에서만 유효하니, 낮은 OS를 사용하는 사용자들을 위해 구현을 해줄 필요가 있다는 것을 알려주는 역할을 한다고 생각했습니다. 그래서 개발의 편의성을 돕는다는 것 이외에 xcode자체에서 기능적인 역할을 한다고 생각하지는 않았습니다. 그리고 말씀해주신 분기처리는 필요성을 느끼고 있지만, 현재 프로젝트의 진행상황을 고려하여, 우선은 앱배포대상을 14.0이상으로 지정해주는 방식으로 변경하였습니다. 분기처리를 해보는 경험도 좋겠지만, 다음 프로젝트에서 진행하게 될 POST메서드를 꼭 구현해보고 싶어서 이렇게 반영하였는데, 괜찮을까요? 🥺🥺

Allie-e replied:
추가적인 리팩토링 과정에서 iOS 14.0 이상에서만 사용할 수 있는 cellRegistration 기능을 register로 변경해주면서 deployment target도 13.0으로 낮춰주었습니다!

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+import UIKit
+
+@available(iOS 14.0, *)
+class ViewController: UIViewController {
```
어떤 역할을 하는 `ViewController`인가요? 새로운 개발자가 팀에 합류하는 상황을 대비해서 어떤 역할을 하는 타입인지 명확한 이름을 지어봅시다!

oksunwoo replied:
앞으로의 이앱에서도 가장 기본이 되고 메인화면이 될 ViewController라고 생각했기에 `MainViewController` 로 이름변경하였습니다! 😇

ryan-son replied:
이 화면이 탭 바 컨트롤러에 속하는 화면이 되고 탭 바를 통해 접근할 수 있는 메인 화면이 여러개가 될 때 이 화면의 이름은 또 바뀌어야 할지 몰라요. 그래서 이 화면이 언제까지나 유일한 메인화면일지는 알 수 없다고 생각한다면 메인보다는 한 단계 더 자세한 이름을 지어주는게 좋을지 모르겠네요.

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+@available(iOS 14.0, *)
+class ViewController: UIViewController {
+    private enum Section: CaseIterable {
+        case main
```
상품이 나타나는 섹션은 언제까지나 `main`이 아닐 수 있지 않을까요? 앞으로의 변화에 유연하게 대응할 수 있도록 해당 섹션을 정확히 지칭할 수 있는 이름을 생각해보는건 어떨까요?

oksunwoo replied:
맞아요 동의합니다. 추가적으로 섹션이 구성될수도 있다고 생각해서, 조금더 구체적으로 `product` 로 변경했습니다! 추후에 섹션을 표기하는 일이 있을때도 main보다는 product라고 하는 것이 좋을 것 같다고 생각했습니다!! 😇

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    private enum Section: CaseIterable {
+        case main
+    }
+    // MARK: - Properties
```
MARK 주석 덕에 가시성이 좋네요! 덤으로 주석이 있는 위치로 이동하기도 편리하군요~! 👏

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    // MARK: - Properties
+    @IBOutlet weak var segment: SegmentedControl!
+    @IBOutlet weak var listCollectionView: UICollectionView!
+    @IBOutlet weak var gridCollectionView: UICollectionView!
```
다른 프로퍼티와 마찬가지로 인터페이스 빌더와 관련된 프로퍼티도 상위 레이어에서 접근하지 못하도록 접근제한자를 설정해봅시다! 키워드의 순서는 어떻게 하는 것이 좋을까요?

oksunwoo replied:
접근제어자 모두 붙여주었습니다. 
키워드 순서는 `@IBOutlet private weak var` 다음과 같은 순서로 작성을 하였는데, private을 weak 앞으로 위치시켜준 이유는 자주 사용되는 키워드를 앞쪽에다가 놓아야한다고 생각했기때문입니다. 혹시 순서에 관련해서 어떠한 방법이 맞는지 여쭤봐도 괜찮을까요? 🤔

ryan-son replied:
이 부분은 스타일이기 때문에 동료와 기준을 만들면 됩니다. 여러분들의 의견도 좋다고 생각해요~

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    @IBOutlet weak var listCollectionView: UICollectionView!
+    @IBOutlet weak var gridCollectionView: UICollectionView!
+    
+    private var productList = [ProductInformation](){
```
공백이 없네요~ 실수인 것 같아요! 😉
```suggestion
    private var productList = [ProductInformation]() {
```

oksunwoo replied:
수정 완료했습니다! 😊

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        didSet {
+            applyListSnapShot()
+            applyGridSnapShot()
+        }
```
데이터의 변화에 UI가 반응형으로 동작할 수 있도록 구성해보셨군요! 👏 멋져요~

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        super.viewDidLoad()
+        
+        listCollectionView.collectionViewLayout = setListCollectionView()
+        gridCollectionView.collectionViewLayout = setGridCollectionView()
```
메서드의 이름을 보면 `setListCollectionView()`은 코드의 동작과는 달리 호출하면 내부에서 `listCollectionView.collectionViewLayout`의 설정을 해줄 것 같네요. 더 좋은 이름은 없을까요? 이름을 두고 사용 방식을 바꿔보는 것도 방법이겠네요. 🤔

oksunwoo replied:
조금 더 구체적으로 `setListCollectionViewLayout`으로 변경해주었습니다! 과거메서드명은 모든 setting을 다 할 것만 같은 느낌이였네요ㅎㅎ 😁

ryan-son replied:
음~ set에는 저는 아래와 같은 느낌이 있다고 생각해요. 여러분은 어떻게 생각하시나요?

```swift
// 1. set~: 호출 시 값을 설정 (세팅)
override func viewDidLoad() {
    super.viewDidLoad()
    ...
    setListCollectionViewLayout()
}

  private func setListCollectionViewLayout() {
      ...
      let layout = UICollectionViewCompositionalLayout(section: section)
      listCollectionView.collectionViewLayout = layout
  }

// 2. make~ or create~: 특정 인스턴스를 생성하여 반환
override func viewDidLoad() {
    super.viewDidLoad()
    ...
    listCollectionView.collectionViewLayout = createListCollectionViewLayout()
}

  private func createListCollectionViewLayout() -> UICollectionViewLayout {
      ...
      let layout = UICollectionViewCompositionalLayout(section: section)
      return layout
  }

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        gridCollectionView.collectionViewLayout = setGridCollectionView()
+        
+        setSegmentedControl()
+        getData()
```
어떤 데이터를 받아오는걸까요? 동일한 ViewController에서 API를 통해 다른 데이터를 받아올 경우를 생각해서 메서드에 더 명확한 이름을 적용하는 것이 좋겠어요~!

Allie-e replied:
상품정보 데이터를 받아온다는 의미로 `getProductData()`로 네이밍 수정해주었습니다! 😊

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    }
+    
+    private func setSegmentedControl() {
+        self.navigationItem.titleView = segment
```
인스턴스의 요소 호출에 다른곳과 달리 `self`를 명시하셨네요. 언제 `self`를 명시적으로 작성할 것인지 동료와 협의해보는건 어떨까요?

Allie-e replied:
self 키워드는 프로퍼티와 일반 변수의 이름이 같을 때 명확한 구분을 위해서 사용한다고 알고 있습니다. 지금 상황에서는 굳이 명시해주지 않아도 될 것 같아요! 꼭 필요한 상황이 아니면 사용하지 않는 쪽으로 맞춰보겠습니다~

ryan-son replied:
인스턴스의 요소임을 명확히 표현하기 위해 self 명시를 선호하는 경우도 있으니 여러분도 선호에 따라 합의하여 결정하시면 됩니다~

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    
+    private func setSegmentedControl() {
+        self.navigationItem.titleView = segment
+        segment.setUpUI()
```
cell과 같이 서브클래싱해서 사용하고 계신데 다른 요소와 달리 외부에서 UI 세팅 요청을 하고 있네요. 일관성 있게 이니셜라이저에서 이 작업을 하는건 어떨까요?

Allie-e replied:
cell 처럼 segmentedControl 클래스 내부에서 `setUpUI()`를 호출해주었습니다! 그런데 이니셜라이저 내부에 작성해주려고 하니 아래와 같이 실패가능한 이니셜라이저를 구현해야하더라구요. 왜 실패가능한 이니셜라이저로 구현해야하는지, 어떤 경우에 실패하는지 궁금합니다ㅠㅠ 잘 찾아봐도 이유에 대한 답은 얻지 못했습니다.. 우선은 객체가 초기화된 후 호출되는 `awakeFromNib()` 메소드를 이용하여 처리해주었는데 이니셜라이저가 아닌 현재 방식을 사용해도 무방한건지도 궁금하네요🤔

```swift
required init?(coder: NSCoder) {
    super.init(coder: coder)
    setUpUI()
}
```

ryan-son replied:
네, 문제 없습니다~ nib의 개념과 `awakeFromNib()`이 어느 시점에 호출되는지 알아보시면 확신을 가지실 수 있으실거에요~

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    }
+    
+    private func getData() {
+        let api = APIManager()
```
ViewController를 테스트할 상황을 고려해서 구체적인 타입을 메서드 내부와 같이 보이지 않는 곳에서 사용하지 않는 것이 좋습니다! 
ViewController를 초기화할 때 `APIManager` 인스턴스를 주입받을 수 있도록 만들어보는건 어떨까요? 덤으로 APIManager의 공개된 인터페이스를 요구사항으로 하는 추상적인 타입을 초기화 시 주입받을 타입으로 하면 완벽해질 것 같아요.

Allie-e replied:
APIManageable이라는 protocol을 구현하여 이 프로토콜을 채택하고 준수하는 어떤 타입이라도 ViewController의 초기화시 들어올 수 있도록 수정해주었습니다 😊 라이언이 어제 해주신 이야기가 많은 도움이 되었어요! 감사합니다!

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            case .failure(let error):
+                print(error)
+            }
+        }
```
향후에 실행될 escaping closure에 ViewController의 인스턴스인 self가 캡처링되고 있네요. **특정한 상황**에서는 이런 경우에 캡처링된 요소가 메모리에서 해제되지 않아 메모리 누수를 일으킬 수 있어요. 아래 내용을 학습해봅시다!

1. 특정한 상황은 어떤 경우일까요?
2. 캡처링된 요소가 메모리에서 해제되지 않는 이유는 무엇인가요?
3. 어떻게 메모리에서 해제될 수 있게 만들어줄 수 있을까요?

oksunwoo replied:
1. `self`와 `closure`가 서로 참조하고 있을때 순환참조가 발생하여 메모리누수를 일으킬 수 있습니다.
2. `closure`가 `self`를 강하게 참조하면서 클로저 실행이 끝나기전까지 `self`는 메모리에서 해제되지 않습니다.
3. `weak` 키워드를 이용하여 순환참조가 발생하지 않도록 해줄 수 있습니다. 추가적으로, `weak`를 사용하는 경우에, `nil`을 할당받을 가능성이 있어서, Optional타입으로 `self`에 대한 바인딩을 해주어야 합니다. 지금과 같은 경우는 ViewController가 해제된 후 불필요한 작업을 피하기위해서 Optional Chaining을 사용하였습니다.

ryan-son replied:
정말 이곳의 escaping closure과 self가 강하게 참조하여 `MainViewController`가 Disappear되더라도 메모리에서 해제되지 않을까요? 디이니셜라이저를 통해 알아보실 수 있습니다. [이 글](https://medium.com/@almalehdev/you-dont-always-need-weak-self-a778bec505ef)도 참고해보세요~

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    private func setUpListCell() {
+        let cellRegistration = UICollectionView.CellRegistration<ListCollectionViewCell, ProductInformation> { cell, indexpath, product in
+            
+            cell.configCell(with: product)
```
config도 충분히 이해할 수 있지만, 더 명확하게 스펠을 생략하지 않고 작성하는 것도 좋아보여요.

Allie-e replied:
습관적으로 줄임말을 사용하고 있었던 것 같아요! 라이언 말씀대로 좀 더 명확히 할 필요가 있다고 생각합니다ㅎㅎ configure로 수정해주었습니다!

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            cell.configCell(with: product)
+        }
+        
+        listDataSource = UICollectionViewDiffableDataSource<Section, ProductInformation>(collectionView: listCollectionView, cellProvider: { (collectionView, indexPath, product) -> ListCollectionViewCell in
```
cellRegistration 초기화 방식과 달리 trailing closure 문법을 사용하지 않으셨네요. 여러분만의 작성 규칙은 무엇인가요~?

oksunwoo replied:
우선, `diffableDataSource`라는 것을 처음 접해보다보니, 예제코드나 여타 블로그 같은 것을 참고하여 사용방식을 익히다보니 작성규칙보단 작성하는법에 중점적으로 기준을 맞추었습니다. 그래서 저희끼리 상의해본 결과, `CellRegistration` 방식을 사용하지 않기로 결정했습니다. `register`로 변경한 핵심 이유는 
1. `cellRegistration` 방식은 ios14버전 이후에서만 사용가능하기에, `register`로 변경시 target 버전을 13으로 낮출수 있는 장점
2. 작성 규칙 준수
입니다. 사실 2번과 같은 경우는 아직 익숙하지 않기에 규칙이 뒤죽박죽하다고 생각이 드는데 라이언의 경우에는 어떠한 작성 규칙을 갖고 계시는지 궁금합니다 😢

ryan-son replied:
저 같은 경우에는 함수가 하나의 클로저만을 받을 때는 trailing closure 방식을 이용하고 (map, filter, reduce 등), `animate(duration:animations:completion:)`과 같이 복수의 클로저를 받는 경우에는 해당 클로저가 어떤 역할을 하는지 명시하기 위해 trailing closure 문법을 사용하지 않습니다.

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            
+            let cell = collectionView.dequeueConfiguredReusableCell(using: cellRegistration, for: indexPath, item: product)
+            
+            return cell
```
줄바꿈 규칙도 궁금합니다~!

Allie-e replied:
리턴해주는 부분에서는 개행을 해주었습니다~! 62번 라인은 실수로 공백이 들어간 것 같아요! 수정하겠습니다!

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    }
+    
+    private func applyListSnapShot(animatingDifferences: Bool = true) {
+        var snapshot = NSDiffableDataSourceSnapshot<Section, ProductInformation>()
```
이 메서드가 호출될 때마다 매번 빈 스냅샷을 초기화하고 있네요. 다음 페이지의 데이터를 수신해서 `productList`에 세팅된다면 기존 상품들은 모두 사라지지 않을까요~?

oksunwoo replied:
아래 댓글에 코멘트 남겼습니다...

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    
+    private func applyListSnapShot(animatingDifferences: Bool = true) {
+        var snapshot = NSDiffableDataSourceSnapshot<Section, ProductInformation>()
+        snapshot.appendSections(Section.allCases)
```
추가 상품이 세팅되어 이 메서드를 다시 호출한다면 이미 추가한 섹션을 추가하려 한다며 런타임 에러가 발생하겠네요. 어떻게 해결해볼까요? 🤔

oksunwoo replied:
```swift
   private func applyListSnapShot(animatingDifferences: Bool = true) {
        guard var snapshot = listDataSource?.snapshot() else {
            return
        }
       
        for section in Section.allCases {
            snapshot.appendItems(productList, toSection: section)
        }
        
        listDataSource?.apply(snapshot, animatingDifferences: animatingDifferences)
    }
```
이러한 방식으로 해결해보려했는데..
<img width="848" alt="스크린샷 2022-01-27 오후 5 23 42" src="https://user-images.githubusercontent.com/55880539/151323206-2db8a3ed-4a6a-4f72-818b-d646f7d3230a.png">

첫번째 섹션을 입력하라고 나오는데, 그렇다고 첫번째 섹션만을 입력해준다면, 모든 product가 첫번째 섹션에만 add가 될 것이라고 생각을 했습니다.. 이부분 역시 미해결 상태이긴한데, 추후에 라이언과 한번 다시 이야기해보는 기회가 있었으면 좋겠습니다. 🥺

ryan-son replied:
네, 같이 이야기해보시죠~ 제가 생각하기에는 최초 1회 dataSource를 구성할 때 section을 추가하고 이후부터는 추가하지 않는 방식으로 가면 될 것 같다고 생각합니다.

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        let item = NSCollectionLayoutItem(layoutSize: itemSize)
+        let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .absolute(view.frame.height * 0.1))
+        let group = NSCollectionLayoutGroup.horizontal(layoutSize: groupSize, subitems: [item])
+        let section = NSCollectionLayoutSection(group: group)
```
이번 기회에 `item`, `group`, `section` 키워드를 잘 이해해 둔다면 CompositionalLayout을 사용하는데 더 자신감을 가질 수 있겠어요.

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        return layout
+    }
+    // MARK: - IBAction Method
+    @IBAction func changeView(_ sender: SegmentedControl) {
```
인터페이스 빌더와 관련된 메서드도 외부로부터 인터페이스를 숨길 수 있어요~

Allie-e replied:
넵 private으로 접근제어자 붙여주었습니다!

ryan-son commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            gridCollectionView.isHidden = true
+        case 1:
+            listCollectionView.isHidden = true
+            gridCollectionView.isHidden = false
```
0 번이 리스트 뷰, 1 번이 그리드 뷰라는 것을 명확히 표현하는 방법을 생각해보는건 어떨까요? 열거형을 활용해서 정의하거나, ViewController의 인스턴스 프로퍼티로 모드를 설정하게하고 모드에 따라 화면을 그리는 등 여러가지 방식을 생각해볼 수 있을 것 같아요.

Allie-e replied:
열거형을 네임스페이스로 활용해서 `ViewMode.list` 와 `ViewMode..grid` 같은 형식으로 표현하도록 수정해주었습니다~!

ryan-son commented on OpenMarket/OpenMarket/Extension/CALayer+extension.swift:
```diff
@@ -0,0 +1,28 @@
+            switch edge {
+            case UIRectEdge.top:
+                border.frame = CGRect.init(x: 0, y: 0, width: frame.width, height: width)
+                break
```
case 내부에서 실행가능한 구문이 있을 경우엔 break를 명시하지 않아도 됩니다~ [Break in a Switch Statement - Swift Programming Language 5.5](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html#ID139)

oksunwoo replied:
네 break 모두 제거해주었습니다 😙

ryan-son commented on OpenMarket/OpenMarket/Extension/CALayer+extension.swift:
```diff
@@ -0,0 +1,28 @@
+import UIKit
+
+extension CALayer {
+    func addBorder(_ arr_edge: [UIRectEdge], color: UIColor, width: CGFloat) -> CALayer {
```
1. 메서드 이름으로 보면 border를 만들어서 subLayer에 추가까지 해줄 것 같은데 동작은 새로운 Layer를 만들어 반환하네요. 지정한 인스턴스를 생성해주는 팩토리 패턴 형식의 메서드니 `makeBorder(_:color:width:)`와 같은 이름도 좋겠어요.
2. 여기에서는 특이하게 단어의 줄임표현과 함께 snake case 형식의 이름을 사용하셨네요. 이유가 있을까요?

oksunwoo replied:
1. 이름 변경했습니다 😁
2. 예제코드 실행시켜보고 미처 수정하지 못했었네요😓 지금 다시 camel case로 변경했습니다.

ryan-son commented on OpenMarket/OpenMarket/Extension/CALayer+extension.swift:
```diff
@@ -0,0 +1,28 @@
+        for edge in arr_edge {
+            switch edge {
+            case UIRectEdge.top:
+                border.frame = CGRect.init(x: 0, y: 0, width: frame.width, height: width)
```
다른 곳과 달리 `.init` 형식으로 인스턴스를 초기화하셨네요. 양식을 일관성 있게 통일하는 것이 좋겠어요~

oksunwoo replied:
네! 지금 .init 형식을 제거해주어 양식을 통일하였습니다 🤩

ryan-son commented on OpenMarket/OpenMarket/Extension/CALayer+extension.swift:
```diff
@@ -0,0 +1,28 @@
+            default:
+                break
+            }
+            border.backgroundColor = color.cgColor;
```
Swift에서 세미콜론은 하나의 라인에 복수의 코드를 작성할 때 사용해요. 여기에서는 필요하지 않아 보이네요!
그리고 이 구문은 위치 상으로 보았을 때 `arr_edge`에 입력되는 요소의 개수만큼 불필요하게 반복해서 동작할 것 같아요. 적당한 위치로 이동시켜 보는건 어떨까요?

oksunwoo replied:
세미콜론 제거하고 ` border.backgroundColor = color.cgColor` 해당 라인 for 문 바깥으로 이동하여 반복 동작 하지 않도록 제어해주었습니다 🙋🏻

ryan-son commented on OpenMarket/OpenMarket/Extension/Double+extension.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+extension Double {
+    func formattedPrice() -> String {
```
동작으로 보았을 때 Price 뿐만 아니라 Double에 해당하는 수라면 모두 구분점을 넣어줄 수 있을 것 같아요. 범용적인 이름으로 바꿔볼 수 있지 않을까요?

oksunwoo replied:
조금 더 범용적으로 `formattedNumber`라는 이름으로 변경해주었습니다👍

ryan-son replied:
해당 타입은 이미 Double인만큼 formatted라는 이름도 괜찮겠네요. 하지만 포매팅한다는 것 자체에는 어떤 형식으로 포매팅한다는 의미가 없으니 문서화 주석이 필요해보여요. 한 번 생각해보셨으니 제 의견을 말씀드리면 `addSeparator()`, `separatorFormatted()`, `withSeparator()` 등으로 지어볼 것 같아요 😊 (절대 현재 상태를 바꾸라는 의미는 아닙니다!)

ryan-son commented on OpenMarket/OpenMarket/Extension/String+extension.swift:
```diff
@@ -0,0 +1,10 @@
+extension String {
+    func strikeThrough() -> NSAttributedString {
+        let attributeString = NSMutableAttributedString(string: self)
+        attributeString.addAttribute(NSAttributedString.Key.strikethroughStyle, value: NSUnderlineStyle.single.rawValue, range: NSMakeRange(0, attributeString.length))
```
range를 작성할 때 Objective-C에서 사용하던 `NSMakeRange(_:_:)`를 구조체 형식으로 래핑하여 멤버와이즈 이니셜라이저를 지원하는 `NSRange(location:length:)`를 이용할 수 있어요. 둘 다 동일하게 동작하지만 Swift로 작성할 때는 Swift 스타일로 작성된 방식을 활용해봅시다~

Allie-e replied:
말씀하신대로 Swift 형식으로 수정해주었습니다! 생각하지 못한 부분이었는데 이런 부분에도 신경쓰도록 노력할게요! 감사합니다ㅎㅎ

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
+        request.httpMethod = httpMethod
+        
+        return request
+    }
```
지정한 형식으로 인스턴스를 만들 수 있도록 팩토리 패턴 형식의 메서드를 작성해보셨네요! 👏 팩토리 패턴 형식의 메서드는  무언가를 만들어 반환한다는 것을 암시할 수 있도록 `make~`, `create~` 등의 동사를 많이 사용합니다.
동일한 내용을 `URLRequest`의 이니셜라이저로도 만들어볼 수 있었을텐데 이 방식을 선택한 이유는 무엇인가요? 
어떤 경우에 팩토리 패턴 형식의 메서드를 작성할지, 이니셜라이저를 작성할지 기준을 정해봅시다~!

Allie-e replied:
이니셜라이저를 사용하면 인스턴스 생성에 대한 세부 사항을 개발자가 일일이 기억하고 있어야한다는 단점이 있을 것 같아요. 현재와 같은 경우에는 설정해주는 사항들이 많지 않지만 기준을 정한다면, 이니셜라이저보다는 팩토리패턴을 사용하는 쪽을 택할 것 같습니다!

ryan-son replied:
이니셜라이저를 활용하든, 팩토리 패턴으로 객체를 만들어 반환하든 각자의 장단점이 있습니다. 만약 `URLRequest`의 `extension`으로 `init(url:httpMethod:)`라는 이니셜라이저를 제공한다면 별다른 내용을 제공하지 않더라도 사용자는 url과 httpMethod를 통해 `URLRequest` 인스턴스를 만들 수 있음을 알 수 있을거에요. 다만 팩토리 메서드는 이름에서 더 많은 내용을 암시할 수 있고, 반환하는 타입에 대한 제약이 이니셜라이저에 비해 자유롭기 때문에 (상속한 타입 중에서 취사선택하여 반환) 유리한면도 있습니다.
다만 현재 여러분께서 작성하신 `request(_:_:)` 메서드는 향후에 여러 타입에서 활용될 여지가 있어보이니 `URLRequest`의 `extension`으로 위치를 조정하는 것도 고려해볼만 하겠네요 (타입 메서드).

```swift
enum HTTPMethod: String {
     case get = "GET"
     case ~
}

extension URLRequest {
    static func make(with url: URL, _ httpMethod: HTTPMethod) -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = httpMethod.rawValue
        return request
    }
}

// 활용
let request = URLRequest.make(with: url, httpMethod: .get)
```

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
-    func requestHealthChecker(completionHandler: @escaping (Result<Data, Error>) -> Void) {
-        guard let url = URL(string: apiHost + "healthChecker") else {
-            completionHandler(.failure(URLSessionError.urlIsNil))
+    private func request(_ url: URL, _ httpMethod: String) -> URLRequest {
```
`httpMethod`를 `String` 보다 미리 만들어둔 `HTTPMethod` 타입의 인스턴스(케이스)를 받을 수 있도록 작성하면 더 편리하게 사용할 수 있을 것 같아요!

oksunwoo replied:
enum 타입도 변경을 해서 HTTPMethod 타입의 인스턴스를 파라미터로 받을 수 있게 변경해주었습니다. 🙆🏼

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
+            guard let statusCode = (response as? HTTPURLResponse)?.statusCode, self.successRange.contains(statusCode) else {
+                completionHandler(.failure(.statusCodeError))
                 return
             }
```
네트워킹에 실패했을 때 어떤 이유로 실패했는지 알 수 있도록 에러에 상태코드를 포함할 수 있게 만들어주면 더 좋을 것 같아요!

Allie-e replied:
Error 타입에 연관값을 활용하여 상태코드를 보여줄 수 있도록 해주었습니다.
기존에 사용하던 successRange.contains(statusCode)를 사용해주고싶어서 Not 논리연산자를 사용해보았는데, 가독성이 좀 떨어진다는 생각이 들어요. 라이언은 어떻게 생각하시는지 궁금합니다! 🤔

```swift
if let statusCode = (response as? HTTPURLResponse)?.statusCode,
       !self.successRange.contains(statusCode) {
       completionHandler(.failure(.statusCodeError(statusCode)))
    }
```

ryan-son replied:
저는 두 가지 방안이 떠오르네요~
1. `notContains(_:)`라는 메서드를 만든다.
2. 경우를 나누어 가독성이 좋은쪽으로 (if, guard) 작성한다 (타입캐스팅 -> statusCode 비교)

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
             completionHandler(.success(data))
         }
         task.resume()
-        
```
`task`를 받아두었으니 이후 취소와 같은 작업에 활용할 수 있도록 반환해주는건 어떨까요?

Allie-e replied:
넵! `performDataTask() `메소드의 리턴 타입으로 URLSessionDataTask를 반환하도록해서 나중에 다른 작업에 task를 활용할 수 있도록 수정해주었습니다~

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
     
     func requestProductInformation(productID: Int, completionHandler: @escaping (Result<ProductInformation, Error>) -> Void) {
-        guard let url = URL(string: apiHost + "/api/products/" + "\(productID)") else {
+        guard let url = URLManager.productInformation(productID).url else {
```
열거형의 associated value를 이용해 세련되게 url을 불러오고 계시네요! 🚀

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
         let task = URLSession.shared.dataTask(with: request) { data, response, error in
-            guard let statusCode = (response as? HTTPURLResponse)?.statusCode, self.successRange.contains(statusCode) else {
-                completionHandler(.failure(URLSessionError.statusCodeError))
+            guard error == nil else {
+                completionHandler(.failure(.requestFail))
                 return
             }
```
이 경우는 언제 발생하나요?

Allie-e replied:
error 부분은 통신 중에 발생한 에러들이 들어온다고 알고 있습니다.
예를들어, 와이파이 연결을 끊으면 요청자체가 서버에 도달하지 못하고, fail되어 error로 들어옵니다. 
실제로 인터넷 연결을 끊어보니, 아래와같이 request가 fail 되는 것을 확인할 수 있었습니다.
![image](https://user-images.githubusercontent.com/83864058/151032544-3827a9bf-089d-4e08-b993-2313f335d1c1.png)

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
-        let pageNumberQuery = URLQueryItem(name: "page_no", value: "\(pageNumber)")
-        let itemsPerPageQuery = URLQueryItem(name: "items_per_page", value: "\(itemsPerPage)")
-        urlComponents?.queryItems?.append(pageNumberQuery)
-        urlComponents?.queryItems?.append(itemsPerPageQuery)
-        
-        guard let url = urlComponents?.url else {
+        guard let url = URLManager.productList(pageNumber, itemsPerPage).url else {
```
URLComponents 타입을 활용하시다 열거형을 활용하는 방식으로 변경하셨네요. 각자 어떤 장단점들이 있었나요?

Allie-e replied:
request를 하는 메소드 안에서 url과 request를 만드는 것은 해당 메소드 내부에서 path나 query를 받아서 urlString을 조합하는 과정까지 하고 있기 때문에 여러 작업을 하고 있다는 생각이 들어 분리해주었습니다. URLComponents를 사용하면 URL 관련 코드를 깔끔하게 작성해서 관리할 수 있다는 장점이 있는 것 같아요. 열거형을 활용하는 방식은 URL을 생성하는 역할을 분리함과 동시에 문자열들을 한곳에서 관리해 유지보수가 쉽다는 장점도 있는 것 같습니다~

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/APIManager.swift:
```diff
@@ -1,79 +1,75 @@
 extension APIManager {
-    func createDataTask<Element: Decodable>(with request: URLRequest, _ completionHandler: @escaping (Result<Element, Error>) -> Void) {
+    func performDataTask<Element: Decodable>(with request: URLRequest, _ completionHandler: @escaping (Result<Element, Error>) -> Void) {
         let task = URLSession.shared.dataTask(with: request) { data, response, error in
```
APIManager가 `URLSession.shared`라는 구체적인 인스턴스에 의존하고 있네요. SOLID 원칙의 Dependency Inversion Principle (DIP)를 위배하고 있어 향후 기능이 추가될 때 Open-Close Principle (OCP)도 준수하기 어려울 수 있겠어요 (기능 확장/수정 시 연관된 메서드를 함께 수정하여야 함). 
이 메서드에 의존하는 다른 메서드들이 유연하게 대응할 수 있도록 사용하는 URLSession의 인스턴스를 외부로부터 주입받아 사용할 수 있는 방법은 없을까요?

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/URLManager.swift:
```diff
@@ -0,0 +1,20 @@
+        case .productInformation(let productID):
+            return URL(string: URLManager.apiHost + "/api/products/" + "\(productID)" )
+        case .productList(let pageNumber, let itemsPerPage):
+            return URL(string: URLManager.apiHost + "/api/products?page_no=\(pageNumber)&items_per_page=\(itemsPerPage)")
```
문자열을 조합할 때 어떤 경우에 +를 사용하고 어떤 경우에 문자열 보간법을 활용하시나요?
이렇게 단일 실행일 경우에는 영향이 없다고 볼 수 있지만 반복적으로 문자열을 조합할 경우에는 문자열 보간법이 가장 빠르게 실행된다고 알려져있어요.

oksunwoo replied:
성능적인 부분을 고려한다면 보간법을 쓰는 것이 좋을 것 같아서, 모두 다 변경해주었습니다 👍 
추가적으로 어떠한 경우에 사용이 되는지 알아보았는데, +는 문자열을 만들때, String의 조합 시 활용가능합니다. 반면, 문자열 보간법은 상수, 변수와 같은 것들을 통해 새로운 문자열을 구성할 수 있습니다. 즉, 먼저 String 타입으로의 변환이 없이, Int와 같은 타입들을 사용하여 문자열을 만들어 낼 수 있다는 장점이 있습니다.

ryan-son replied:
문자열 조합을 위해 굉장히 복잡한 로직을 구성하지 않고서야 큰 차이를 체감하기는 어려울 수 있어요. 본인만의 기준이 있는 것이 가장 중요합니다!

ryan-son commented on OpenMarket/OpenMarket/View/Cell/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,143 @@
+        configUI()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
구현하지 않은 인터페이스는 avilable Declaration attribute를 이용해 다른 사용자들이 활용할 수 없도록 만드는 것이 좋을 수 있겠어요!

oksunwoo replied:
```swift
   @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
```
다음과 같이 구현하지 않은 인터페이스에 키워드를 붙여주었습니다!

ryan-son commented on OpenMarket/OpenMarket/View/Cell/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,143 @@
+    
+    private func setUpImage(url: String) {
+        guard let imageUrl = URL(string: url),
+              let data = try? Data(contentsOf: imageUrl) else {
```
스크롤링 하실 때 버벅임이 느껴지시지 않으셨나요? 현재는 이 메서드가 호출될 때마다 동기적으로 작동하는 메서드인 `Data(contentsOf:)`가 메인스레드에서 실행되고 있어 UI 스레드 블로킹을 유발하고 있습니다. [이 내용](https://developer.apple.com/documentation/foundation/nsdata/1413892-init)을 살펴보시고 이 부분의 처리를 어떤 방식으로 변경하면 좋을지 고민해봅시다~!

![image](https://user-images.githubusercontent.com/69730931/150636520-2edf1ab8-97e7-4daa-ba05-e3534ff50c00.png)

oksunwoo replied:
메서드가 동기적으로 호출되기에 해당 부분을 global 스레드로 옮겨서 작업하도록 구현해보았습니다. 확실히 스크롤링이 부드러워진 것을 체감할 수 있었는데요. 이렇게 구현을 해도 될까요?

```swift
    private func setUpImage(url: String) {
        DispatchQueue.global().async {
            guard let imageUrl = URL(string: url),
                  let data = try? Data(contentsOf: imageUrl) else {
                return
            }
            
            DispatchQueue.main.async {
                self.productImageView.image = UIImage(data: data)
            }
        }
    }
```
그리고 저희가 문제를 하나 발견했는데, 스크롤을 하면 이미지가 이전에 있던 이미지로부터 차례로 변경되는 현상을 발견했습니다. 셀이 재사용되면서 가지고있던 모든 이미지를 보여준다고 판단을 하고, 그 문제를 해결하기 위해 `indexPath`랑 `collectionView`의 `indexPath`가 일치할 경우에만 이미지를 출력하도록 작성하다가 잘 해결되지 않아 문제가 발생한 것은 그대로 남아있는 상태입니다..😓

oksunwoo replied:
```swift

    private func setUpImage(url: String) {
        DispatchQueue.global().async {
            guard let imageUrl = URL(string: url) else {
                return
            }
            
            self.loadImage(url: imageUrl) { image in
                DispatchQueue.main.async {
                    self.productImageView.image = image
                }
                
            }
        }
    }
    
    private func loadImage(url: URL, completionHandler: @escaping (UIImage) -> Void) {
        let task = URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else {
                return
            }
            
            guard let image = UIImage(data: data) else {
                return
            }
            
            completionHandler(image)
        }
        task.resume()
    }
```
이러한 방식으로 바꾸어주었습니다! 다만, 이미지가 변경되는 현상은 여전한 것 같습니다😓

ryan-son replied:
여러분들이 알아보셨듯 `Data(contentsOf:)` 메서드는 네트워킹에 적합한 메서드는 아닙니다. 그래서 별도의 스레드 처리를 해주어야 하는 것이죠. 그보다 공식문서에서 밝히고 있듯이 기본적으로 비동기를 지원하는 `URLSession`의 `dataTask`를 통해 이 기능을 구현한다면 더 안정적으로 구현할 수 있을 거에요. 거기에 더해 `URLSessionDataTask`는 작업 취소를 위한 인터페이스를 제공하니 이를 이용해서 원치 않는 작업을 취소한다면 썸네일이 변경되는 문제를 해소할 수 있을 것 같네요.

ryan-son commented on OpenMarket/OpenMarket/View/Cell/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,177 @@
+        accessoryImageView.image = UIImage(systemName: "chevron.right")
+        accessoryImageView.tintColor = .systemGray
+    }
+ 
+    private func setUpImage(url: String) {
+        guard let imageUrl = URL(string: url),
+              let data = try? Data(contentsOf: imageUrl) else {
+                  return
+              }
+        
+        productImageView.image = UIImage(data: data)
+    }
```
여기도 비동기 처리를 해주어야 하겠네요~

oksunwoo replied:
비동기 처리 완료해주었습니다!

ryan-son commented on OpenMarket/OpenMarket/View/SegmentedControl.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+class SegmentedControl: UISegmentedControl {
```
일반적인 이름을 사용하고 있네요. 사용하는 입장에서 어떤 `SegmentedControl`인지 인지할 수 있도록 이름을 변경해보는건 어떨까요?

Allie-e replied:
SegmentedControl에 따라서 List와 Grid 두 타입의 뷰를 보여주는 역할이라서 Layout SegmentedControl 정도로 변경해보았습니다!

ryan-son commented on OpenMarket/OpenMarket/View/Cell/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,177 @@
+            accessoryImageView.widthAnchor.constraint(equalTo: accessoryImageView.heightAnchor)
+        ])
+        
+        accessoryImageView.image = UIImage(systemName: "chevron.right")
```
자주 사용하는 디자인 요소가 있다면 어딘가에 모아두고 호출해서 사용해서 실수를 줄일 수 있겠어요.

Allie-e replied:
AccessoryImage 열거형을 만들어 자주사용하는 image들은 따로 관리하도록 해주었습니다! 라이언이 말씀하신 방향이 맞을까요?

ryan-son reviewed (changes requested):
앨리, 토니, 네트워킹과 Collection View 등 많은 내용을 공부하시고 적용해보시느라 고생 많으셨습니다! 👏

전반적으로 코드가 구조화되어 있어 읽기에 편했고, 각 타입의 역할을 잘 분리하려고 한 노력이 느껴졌습니다. 더불어 많은 시행착오를 겪어보시고 관련된 내용을 고민해보셔서 단기간에 많은 성장을 이루었다고 느낍니다!

> # segmentedControl을 통한 뷰 전환
> 1. **2개의 view를 한 화면에 동시에 띄우고, isHidden으로 한 view를 감추는 방식**
> 2. **view를 하나만 사용하고, 화면전환시 cell의 레이아웃들을 변경하는 방식**

1 번 방식은 여러분이 느끼셨듯이 `listCollectionView`와 `gridCollectionView`의 스크롤 범위에 (`contentOffset`) 차이가 있어 상대적으로 스크롤 범위가 적을 수 밖에 없는 `listCollectionView` 내부의 cell들 사이에 빈 공간이 생기는 현상이 발생될 수 있습니다. `gridCollectionView`가 숨겨진 상태일 때는 화면에 보이지만 않을 뿐 그 공간을 차지하고 있어 그렇습니다. 이 문제를 해결하려면 UI 요소가 숨겨져있을 때 동적으로 크기가 조정되는 `UIStackView` 내부에 collectionView들을 넣는 방식을 고려해볼 수 있겠네요.
또한, 1 번 방식은 받아오는 데이터가 크다면 segmentedControl을 통해 셀 형식이 전환될 때 사용자 경험에서의 이점이 있을 수 있지만, 그만큼 초기 로드 시간이 길어진다는 단점이 있을 것으로 예상됩니다. 더욱이 현재 캐싱을 활용하고 있지 않기 때문에 같은 이미지를 listCell과 gridCell이 두 번씩 요청하여 스크롤 시 반응성이 좋지 않을 수 있습니다 (현재 메인스레드에서 이미지 로드 작업 중).

2 번 방식은 필요한 데이터만 로드하고, 필요한 cell만 dequeue 할 수 있다는 장점이 있습니다. 반면 앞서 언급하셨듯이 segmentedControl을 통해 화면 전환 시 초기 로드 시간이 필요합니다. 이 문제를 캐싱을 통해 어느정도 해소할 수 있을 것으로 판단되네요.

말씀하셨듯이 정답은 없으니 선호하시는 방식으로 구현하시면 되겠습니다.

> defaultContentConfiguration() 사용에 대한 고민

말씀하셨듯이 `defaultContentConfiguration()`으로는 높은 자유도의 커스텀을 수행하기에는 어려움이 있습니다. 이런 경우에는 커스텀 Cell을 만들어 사용하는 것이 유효하다 생각합니다. `defaultContentConfiguration()`을 활용하면 좋은 경우는 기존 `UITableViewCell`에서 제공했던 textLabel, detailedTextLabel 등만을 활용할 때라 생각합니다. Apple에서 기본적으로 제공하는 셀 스타일인 plain, grouped, insetGrouped, sidebar, sidebarPlain을 통해 기존 보다 더 다양한 기본 스타일을 제공하여 활용도가 더 높아졌다 생각합니다.

> 아래와 같이 image 최대사이즈를 제한해 줄 수는 있는 것 같은데, image 사이즈를 정해주는 다른 방법이 있는지 궁금합니다! 🤔

[`reservedLayoutSize`](https://developer.apple.com/documentation/uikit/uilistcontentconfiguration/imageproperties/3601000-reservedlayoutsize)가 말씀하신 역할을 수행할 것으로 예상되네요. 

>## 💢 해결이 되지 않은 점

>### ListCell의 실종사건👮🏻

이 내용은 앞서 언급했으니 넘어가겠습니다~

> AccessoryView의 실종사건👮🏻‍♀️

`AccessoryView`는 Cell이 최초에 인스턴스화될 때 `configUI()`를 통해 이미지가 설정되는 것으로 보이네요. 만들어 둔 셀을 재활용하는 과정에서 기존에 있던 이미지를 없앤다면 명령한대로 없어지는 것이 맞지 않을까요? Cell은 재활용되니 이니셜라이즈 되지 않을 것이고 `configUI()`도 다시 호출하지 않으니 이미지가 지워진 상태가 유지되겠네요. 이 내용에 대해 더 자세히 알아보시고 싶으시면 이니셜라이즈 시점과 재사용 시점에 출력문을 넣고 실행해보시면 되겠습니다.

> ## ❓ 질문 드리고 싶은 부분

> ### Cell을 재사용할 때, attributedText가 reset되지 않는 문제

`UIKit`의 소스는 공개되지 않아 왜 이런 현상이 발생하는지 알 수는 없을 것 같아요. [커뮤니티](https://stackoverflow.com/questions/58628067/cant-reset-uilabel-attributedtext-when-a-uitableviewcell-is-reused)에서도 이유는 밝혀지지 않았지만 `attributedString`을 먼저 할당 해제하면 문제가 해결된다는 내용만 논의가 이루어진 것으로 보여요.

> Cell의 이중등록?

저는 아래 방식으로 등록했는데 문제가 발생하지 않네요. 디스코드에서 이야기 나누어보도록 합시다~!

```swift
private func setUpListCell() {
    listCollectionView.register(ListCollectionViewCell.self, forCellWithReuseIdentifier: "ListCollectionViewCell")
    
    listDataSource = UICollectionViewDiffableDataSource<Section, ProductInformation>(collectionView: listCollectionView, cellProvider: { (collectionView, indexPath, product) -> ListCollectionViewCell in
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "ListCollectionViewCell", for: indexPath) as? ListCollectionViewCell else {
            return ListCollectionViewCell()
        }
        cell.configCell(with: product)
        
        return cell
    })
}

private func setUpGridCell() {
    gridCollectionView.register(GridCollectionViewCell.self, forCellWithReuseIdentifier: "GridCollectionViewCell")
    
    gridDataSource = UICollectionViewDiffableDataSource<Section, ProductInformation>(collectionView: gridCollectionView, cellProvider: { (collectionView, indexPath, product) -> GridCollectionViewCell in
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "GridCollectionViewCell", for: indexPath) as? GridCollectionViewCell else {
            return GridCollectionViewCell()
        }
        cell.configCell(with: product)
        return cell
    })
}
```

> 레이아웃 관련 로그

일단 CompositionalLayout을 통해 셀 크기를 지정해두었으니 기기의 화면크기에 맞추어 셀 크기가 결정되겠네요. 그렇다면 동적으로 셀 높이를 만들지는 않아도 되니 위치와 크기만 잘 잡으면 되겠어요. 

1. `UIStackView` 내부에서 지정된 방식으로 자동으로 레이아웃 되기를 원하는 요소는 `translatesAutoresizingMaskIntoConstraints` 옵션을 끄지 말아주세요. 오토레이아웃을 지정하지 않을 요소는 AutoresizingMask를 통해 레이아웃이 되도록 만들어주어야 합니다. `nameStackView`나 `priceStackView`와 같은 것들이 해당되겠네요.
2. imageView와 같이 UIStackView 내부에서 크기와 위치 관계를 잡기 어려운 요소는 `StackView` 바깥으로 빼서 별도로 레이아웃을 잡는 것도 도움이 됩니다.

궁금한 점이 있으시면 언제든지 물어보세요~ 화이팅! 💪

Allie-e committed:
refactor: storyboard의 사용하지 않는 프로토타입 cell 삭제 #6

oksunwoo committed:
refactor: target deployment 변경 #6

Allie-e committed:
refactor: 모호한 네이밍 수정 #6

oksunwoo committed:
refactor: Interface Builder 관련 접근제어자 설정 #6

Allie-e committed:
refactor: 필요없는 self 키워드 삭제 #6

oksunwoo committed:
refactor: segmentControl의 UI 세팅 요청 변경 #6

Allie-e committed:
refactor: APIManageable protocol 구현 #6

- ViewController 초기화시 APIManager 인스턴스를 주입받도록 init 구현
- APIManageable을 준수하는 어떤 타입이라도 init할 때 들어올 수 있도록 구현

oksunwoo committed:
refactor: escaping closure 내부 순환 참조 방지 #6

Allie-e committed:
refactor: cell register 방식 변경 #6

- deployment target iOS 13.0 으로 변경

oksunwoo committed:
refactor: 열거형을 사용하여 뷰의 전환시 case를 명확히 표현 #6

Allie-e committed:
refactor: extension CALayer 메소드 네이밍 및 로직 수정 #6

- 팩토리 패턴 형식에 맞게 네이밍 수정
- 불필요하게 반복되는 코드 위치 수정

oksunwoo committed:
refactor: NSMakeRange를 NSRange로 수정 #6

Allie-e committed:
refactor: APIManageable 함수 팩토리 패턴에 맞는 네이밍 수정 #6

oksunwoo committed:
refactor: 파라미터를 HTTPMethod 타입의 인스턴스를 받도록 변경 #6

Allie-e committed:
refactor: API request error 메세지에 statusCode 추가 #6

oksunwoo committed:
refactor: task를 반환하여 다른작업에서도 활용할 수 있도록 구현 #6

Allie-e committed:
refactor: URLManager 문자열 조합 방식 보간법으로 변경 #6

oksunwoo committed:
refactor: available Declaration attribute 부여 #6

Allie-e committed:
refactor: AccessoryImage 관리를 위한 열거형 구현 #6

oksunwoo committed:
refactor: 스크롤링 버벅임 개선 #6

- 메인스레드에서 실행되는 메서드를 global 스레드에서 비동기로 실행
- image를 넣는 UI작업을 메인스레드에서 실행

oksunwoo committed:
refactor: 이미지를 호출하는 방식 변경 #6

oksunwoo committed:
refactor: GridCell의 이미지 세팅 방식 변경 #6

ryan-son commented:
이번 스텝의 요구사항을 충분히 만족했다고 판단하여 머지합니다. 고생많으셨습니다! 👏

추가로 대화 나누고 싶거나 전달하고 싶은 내용을 여러분이 남겨주신 코멘트에 남겨두었으니 한 번 읽어보시길 바랍니다!
얼마 남지 않았지만 다음 스텝도 경험해보시고 궁금하신 점은 질문해주시길 바라요! 😊

ryan-son commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,141 @@
+        getProductData()
+        setUpListCell()
+        setUpGridCell()
+        applyListSnapShot(animatingDifferences: false)
+        applyGridSnapShot(animatingDifferences: false)
```
통신이 완료되어 productList가 세팅되면 프로퍼티 옵저버에 위치한 메서드가 실행되지 않나요? 어떻게 보면 불필요할 수 있겠네요.

ryan-son commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,141 @@
+        applyGridSnapShot(animatingDifferences: false)
+    }
+    
+    private func setSegmentedControl() {
+        navigationItem.titleView = segment
+    }
```
이런 방식도 이용해볼 수 있어요~ (이렇게 바꾸라는건 아니에요)

```swift
@IBOutlet private weak var segment: LayoutSegmentedControl! {
    didSet {
        navigationItem.titleView = segment
    }
}
```
`segment` 보다 더 좋은 이름도 생각해볼까요?

ryan-son commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,141 @@
+    }
+    
+    private func applyListSnapShot(animatingDifferences: Bool = true) {
+        var snapshot = NSDiffableDataSourceSnapshot<Section, ProductInformation>()
+        snapshot.appendSections(Section.allCases)
```
이야기 나눈 부분이네요. 다음 스텝에서 고쳐봅시다~

ryan-son commented on OpenMarket/OpenMarket/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,141 @@
+    private func setUpGridCell() {
+        gridCollectionView.register(GridCollectionViewCell.self, forCellWithReuseIdentifier: "GridCollectionViewCell")
+        
+        gridDataSource = UICollectionViewDiffableDataSource<Section, ProductInformation>(collectionView: gridCollectionView, cellProvider: { (collectionView, indexPath, product) -> GridCollectionViewCell in
```
라인이 길어지면 어디에서 줄바꿈할지 기준에 대해 생각해보신적 있으신가요~? 현재 프로젝트에서는 적용하고 있지 않으니 지금부터 적용하실 필요는 없을 것 같고, 생각만 해두시다 이후에 이 내용에 관심있는 동료와 함께하게 된다면 자신만의 기준에 대해 이야기 나누어보는 것도 좋아보이네요.

ryan-son commented on OpenMarket/OpenMarket/Extension/CALayer+extension.swift:
```diff
@@ -0,0 +1,24 @@
+import UIKit
+
+extension CALayer {
+    func makeBorder(_ edges: [UIRectEdge], color: UIColor, width: CGFloat) -> CALayer {
```
세팅이 아니라 새 보더를 만드는거라면 이 메서드가 인스턴스 메서드일 필요는 없지 않을까요? 오히려 어색해 보일 수 있을 것 같아요.

ryan-son commented on OpenMarket/OpenMarket/Model/Networking/URLManager.swift:
```diff
@@ -0,0 +1,20 @@
+    private static let apiHost = "https://market-training.yagom-academy.kr/"
+    
+    case healthChecker
+    case productInformation(Int)
+    case productList(Int, Int)
```
associated value에 어떤 내용이 들어가는지 주석이나 파라미터 이름으로 표현해둔다면 더 좋을 것 같아요.

ryan-son commented on OpenMarket/OpenMarket/View/AccessoryImage.swift:
```diff
@@ -0,0 +1,5 @@
+import UIKit
```
UIKit은 어차피 최소 한 곳에서 모든 프레임워크를 불러다 사용하니 별 의미 없을 수 있겠지만, 아래와 같이 import하는 것도 가능해요 (잡지식)

```swift
import UIKit.UIImage
```

ryan-son commented on OpenMarket/OpenMarket/View/AccessoryImage.swift:
```diff
@@ -0,0 +1,5 @@
+import UIKit
+
+enum AccessoryImage {
+    static let chevron = UIImage(systemName: "chevron.right")
+}
```
앱의 규모에 따라 다르겠지만 Asset -> imageChevronRight 이런 형식으로 관리하는 것도 가능하겠네요. 디자인시스템이 마련되어 자주 사용하는 색상이나 이미지가 있어도 이런식으로 명명규칙을 만들어두면 혼동할 일도 적을 것이고, 무엇보다도 UI 코드 내부에서 String literal을 통해 이미지를 불러오지 않아 이로 인한 문제는 발생하지 않을 것 같아요.

ryan-son commented on OpenMarket/OpenMarket/View/Cell/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+    
+    private func loadImage(url: URL, completionHandler: @escaping (UIImage) -> Void) {
+        let task = URLSession.shared.dataTask(with: url) { data, _, _ in
+            guard let data = data else {
+                return
+            }
+            
+            guard let image = UIImage(data: data) else {
+                return
+            }
+            
+            completionHandler(image)
+        }
+        task.resume()
+    }
```
네트워크를 통해 이미지를 전문으로 가져오는 역할을 수행하는 친구를 만들어도 좋겠어요. UI 요소인 Cell이 네트워킹까지 다루기엔 역할이 많아보일 수 있겠어요.

ryan-son commented on OpenMarket/OpenMarket/View/Cell/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,198 @@
+    }
+    
+    private func setUpLabelStackView() {
+        labelStackView.axis = .vertical
+        labelStackView.alignment = .fill
+        labelStackView.distribution = .fillEqually
+        labelStackView.translatesAutoresizingMaskIntoConstraints = false
```
알고계시겠지만, 이런 방식으로 생성할 당시에 프로퍼티를 초기화할 수 있으니 참고하세요~

```swift
private let labelStackView: UIStackView = {
    let stackView = UIStackView()
    stackView.axis = .vertical
    stackView.alignment = .fill
    stackView.distribution = .fillEqually
    stackView.translatesAutoresizingMaskIntoConstraints = false
    return stackView
}()
```

이런 번거로운 작업을 간단하게 해주는 [Then](https://github.com/devxoul/Then)이라는 편의성 라이브러리도 있어요.
```swift
// Then을 활용한 초기화 시 프로퍼티 설정
private let labelStackView = UIStackView().then {
    $0.axis = .vertical
    $0.alignment = .fill
    $0.distribution = .fillEqually
    $0.translatesAutoresizingMaskIntoConstraints = false
}

// 인스턴스 생성 후 프로퍼티 설정
productNameLabel.do {
    $0.text = data.name
    $0.font = .preferredFont(forTextStyle: .headline)
}
```

ryan-son commented on OpenMarket/OpenMarket/View/LayoutSegmentedControl.swift:
```diff
@@ -0,0 +1,23 @@
+        setUpUI()
+    }
+    
+    func setUpUI() {
```
외부에서 활용하지 않는다면~?

ryan-son reviewed (commented):
추가 코멘트 남깁니다~!
